cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java

```diff
@@ -106,23 +108,39 @@ public void put(Job job) {
   public Job transitState(Job job, Job.State state) {
     // assert that the job state transition is valid
     job.getState().checkTransition(state);
-    Job newJob = new SimpleJob(job.getSchedule(), job.getCreationTime(), job.getNotifications(), state);
+    Job newJob = new SimpleJob(job.getSchedule(), job.getCreationTime(), job.getNotifications(), state,
+                               job.getScheduleModificationTime());
     put(newJob);
     return newJob;
   }
 
   @Override
-  public void addNotification(ProgramSchedule schedule, Notification notification) {
+  public void addNotification(ProgramScheduleRecord record, Notification notification) {
     boolean jobExists = false;
+    ProgramSchedule schedule = record.getSchedule();
     try (CloseableIterator<Job> jobs = getJobsForSchedule(schedule.getScheduleId())) {
       while (jobs.hasNext()) {
-        jobExists = true;
         Job job = jobs.next();
         if (job.getState() == Job.State.PENDING_TRIGGER) {
           // only update the job's notifications if it is in PENDING_TRIGGER, so as to avoid conflict with the
           // ConstraintCheckerService
-          addNotification(job, notification);
-          break;
+          if (job.isToBeDeleted()) {
+            // ignore, it will be deleted by ConstraintCheckerService
+            continue;
+          }
+          long scheduleLastUpdated = record.getMeta().getLastUpdated();
+          if (job.getScheduleModificationTime() != scheduleLastUpdated) {
+            // schedule has changed: this job is obsolete
+            // this might get a conflict (small chance)
+            table.put(getRowKey(job.getJobKey().getScheduleId(), job.getJobKey().getCreationTime()),
+                      TO_DELETE_COL, Bytes.toBytes(scheduleLastUpdated));
```

Comment:
`scheduleLastUpdated ` should be replaced with current time. Maybe we can use a different column called `EXPIRED_COL` to mark the current time, so that it won't have conflict with `markJobsForDeletion` method. In the future, timed out jobs can also be marked in this `EXPIRED_COL` to avoid conflict.